pkg/query/storeset.go

```diff
@@ -311,6 +332,57 @@ func externalLabelsFromStore(st *storeRef) string {
 	return tsdbLabels.String()
 }
 
+func (s *StoreSet) updateHealthyStoreStatus(sr *storeRef, msg string) {
+	s.storesStatusesMtx.Lock()
+	defer s.storesStatusesMtx.Unlock()
+
+	s.storeStatuses[sr.addr] = &StoreStatus{
+		Name:        sr.addr,
+		Labels:      sr.labels,
+		Healthy:     true,
+		LastCheck:   time.Now(),
+		LastMessage: msg,
+		MinTime:     sr.minTime,
+		MaxTime:     sr.maxTime,
+	}
+}
+
+func (s *StoreSet) updateUnhealthyStoreStatus(sr *storeRef, err error) {
+	s.storesStatusesMtx.Lock()
+	defer s.storesStatusesMtx.Unlock()
+
+	minTime, maxTime := int64(-1), int64(-1)
+	if err == nil {
```

Comment:
Can you instead do this (before mutex lock/unlock)?
```
if err == nil {
  warningLogFunction("updateUnhealthyStoreStatus was called with nil error, stack trace: %s", string(debug.Stack()))
  return
}
```
, because if error is nil, the `LastMessage: err.Error()` is going to throw null pointer exception. It's very unlikely that err is ever going to be nil, but little bit of defensive programming can't hurt us.

Reply:
@stefan-improbable is right `err` needs to be filtered out, otherwise panic will occur.

With the comment above with excluding any status store message you can actually make just one `updateStoreStatus` that gets err and check if it's nil or not inside here to tell if error should be passed and if healthy should be true or false (in fact `Healthy` can be dropped)

So you limit the case that someone will use `updateUnhealthyStoreStatus` with err == nil